### PR TITLE
powdr-asmopt: remove unused submachines, instructions, registers

### PR DESCRIPTION
**Pull request by gzanitti**
Solves #682

### CODE REVIEW

**Comment by gzanitti** [comment]
If necessary, we could separate this PR into three smaller parts: 
1) The new ReferencedSymbols implementations.
2) The asmopt module
3) The integration of asmopt into the pipeline

File: asmopt/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+
+const MAIN_MACHINE_STR: &str = "::Main";
+const PC_REGISTER: &str = "pc";
+
+pub fn optimize(mut analyzed_asm: AnalysisASMFile) -> AnalysisASMFile {
+    // Optimizations assume the existence of a Main machine as an entry point.
+    // If it doesn't exist, return the ASM as-is to prevent removing all machines,
+    // which would break some examples from the book.
+    let main_machine_path = parse_absolute_path(MAIN_MACHINE_STR);
+    if analyzed_asm
```
Comment (gzanitti) [inline asmopt/src/lib.rs:21]: Main seems to be the entry point for execution, but it is not mandatory to include it. This check avoids optimizing cases where the Main machine does not exist, thus avoiding that all machines are eliminated (cases like `pil_at_module_level.asm`).

File: asmopt/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+}
+
+fn machine_remove_unused_instructions(machine: &mut Machine, symbols: &HashSet<String>) {
+    machine
+        .instructions
+        .retain(|ins| symbols.contains(&ins.name));
+}
+
+fn collect_constrained_machines(asm_file: &AnalysisASMFile) -> impl Iterator<Item = String> + '_ {
+    asm_file
```
Comment (gzanitti) [inline asmopt/src/lib.rs:127]: This function searches for constrained machines by looking for machines where columns are defined, but perhaps there is a more precise way?

File: asmopt/tests/optimizer.rs [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+
+    machine Unused with degree: 8 {
+        reg pc[@pc];
+        col witness w;
+        w = w * w;
+    }
+    "#;
+
+    let expectation = r#"machine Main with degree: 8 {
+    ::Required required
```
Comment (pacheco) [inline asmopt/tests/optimizer.rs:220]: Shouldn't `Required` be removed here? It is not used anymore

File: asmopt/tests/optimizer.rs [hunk 4]
```diff
@@ -0,0 +1,311 @@
+use powdr_analysis::analyze;
+use powdr_asmopt::optimize;
+use powdr_parser::parse_asm;
+
```
Comment (pacheco) [inline asmopt/tests/optimizer.rs:4]: test suggestion: machines in `link`s (outside of instructions) should not be removed

File: ast/src/asm_analysis/mod.rs [hunk 5]
```diff
@@ -882,6 +882,17 @@ impl Module {
         self.ordering.push(StatementReference::Module(name));
     }
 
+    pub fn retain_machines(&mut self, names: HashSet<String>) {
```
Comment (pacheco) [inline ast/src/asm_analysis/mod.rs:885]: is it clear from the context what `String` should be here?

File: pilopt/src/referenced_symbols.rs [hunk 6]
```diff
@@ -149,8 +167,226 @@ fn symbols_in_expression(
     }
 }
 
-impl ReferencedSymbols for Type {
+fn symbols_in_expression_asm(
```
Comment (pacheco) [inline pilopt/src/referenced_symbols.rs:170]: should this be a `symbols()` method in `ExpressionAsm` objects? maybe implement the `ReferencedSymbols` trait?

File: test_data/asm/book/declarations.asm [hunk 7]
```diff
@@ -5,9 +5,8 @@ mod utils {
         0 => 0,
         _ => arr[len - 1] + sum(len - 1, arr)
     };
-    // A simple function that returns the input incremented by one,
-    // as an expression.
-    let incremented: expr -> expr = |x| x + 1;
+    // A simple function that returns the same input as an expression.
```
Comment (pacheco) [inline test_data/asm/book/declarations.asm:8]: are these book changes related?

File: pipeline/src/pipeline.rs [hunk 8]
```diff
@@ -773,6 +776,7 @@ impl<T: FieldElement> Pipeline<T> {
 
                 self.log("Run analysis");
                 let analyzed_asm = powdr_analysis::analyze(resolved)?;
+                self.maybe_write_pil(&analyzed_asm, "_asm")?;
```
Comment (pacheco) [inline pipeline/src/pipeline.rs:779]: won't this generate a `foo_asm.pil`?

File: pipeline/src/pipeline.rs [hunk 9]
```diff
@@ -787,14 +791,35 @@ impl<T: FieldElement> Pipeline<T> {
         Ok(self.artifact.analyzed_asm.as_ref().unwrap())
     }
 
+    pub fn compute_optimized_asm(&mut self) -> Result<AnalysisASMFile, Vec<String>> {
```
Comment (pacheco) [inline pipeline/src/pipeline.rs:794]: i think this should return a ref to `AnalysisAsmFile` instead of a clone

File: pipeline/src/pipeline.rs [hunk 10]
```diff
@@ -787,14 +791,35 @@ impl<T: FieldElement> Pipeline<T> {
...
+
     pub fn compute_constrained_machine_collection(
         &mut self,
     ) -> Result<&AnalysisASMFile, Vec<String>> {
         if self.artifact.constrained_machine_collection.is_none() {
             self.artifact.constrained_machine_collection = Some({
-                self.compute_analyzed_asm()?;
-                let analyzed_asm = self.artifact.analyzed_asm.take().unwrap();
-                powdr_asm_to_pil::compile::<T>(analyzed_asm)
+                self.compute_optimized_asm()?;
```
Comment (pacheco) [inline pipeline/src/pipeline.rs:820]: this already returns the `optimized_asm`

File: asmopt/tests/optimizer.rs (hunk 3 above)
Comment (gzanitti) [reply asmopt/tests/optimizer.rs:220]: In fact, it's being used, it just doesn't print (I'm going to open an issue on this). That's partly why I created the constrained_machines that are no longer needed (because they really weren't).

File: pilopt/src/referenced_symbols.rs (hunk 6 above)
Comment (gzanitti) [reply pilopt/src/referenced_symbols.rs:170]: It's also implemented. But since symbols() uses all_children this function is used to get “first level” PublicReferences/References or nothing, without checking their children.

File: pipeline/src/pipeline.rs (hunk 10 above)
Comment (gzanitti) [reply pipeline/src/pipeline.rs:817]: Yes, but I understand that this way you avoid having to clone the file. It's similar to how it's done for the PIL optimization pass.

File: asmopt/tests/optimizer.rs (hunk 4 above)
Comment (gzanitti) [reply asmopt/tests/optimizer.rs:4]: Added, thanks!

File: test_data/asm/book/declarations.asm (hunk 7 above)
Comment (pacheco) [reply test_data/asm/book/declarations.asm:8]: minor, but why not keep the `incr` instruction and call it from `main`? it is a bit more "interesting" of an example

File: asmopt/tests/optimizer.rs (hunk 3 above)
Comment (pacheco) [reply asmopt/tests/optimizer.rs:220]: mmm it's required because its a parameter to the submachine?

Comment (gzanitti) [reply asmopt/tests/optimizer.rs:220]: Oh, sorry, yes, exactly. Correctly printed, this should look like this:

```
::Required required
::ParamMachine sub(required)
```

**Review by pacheco — APPROVED** [review]
lgtm
